go/libraries/doltcore/sqle/lookup/range.go

```diff
@@ -0,0 +1,273 @@
+// Copyright 2020 Dolthub, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package lookup
+
+import (
+	"github.com/dolthub/dolt/go/libraries/doltcore/table/typed/noms"
+	"github.com/dolthub/dolt/go/store/types"
+)
+
+// Range represents the contiguous set of values that a lookup operation covers.
+type Range struct {
+	LowerBound Cut
+	UpperBound Cut
+}
+
+var (
+	// alwaysContinueRangeCheck will allow the range to continue until the end is reached.
+	alwaysContinueRangeCheck noms.InRangeCheck = func(tuple types.Tuple) (bool, error) {
+		return true, nil
+	}
+	// neverContinueRangeCheck will immediately end.
+	neverContinueRangeCheck noms.InRangeCheck = func(tuple types.Tuple) (bool, error) {
+		return false, nil
+	}
+)
+
+// OpenRange returns a range representing {l < x < u}.
+func OpenRange(lower, upper types.Tuple) Range {
+	return Range{
+		Above{key: lower},
+		Below{key: upper},
+	}
+}
+
+// ClosedRange returns a range representing {l <= x <= u}.
+func ClosedRange(lower, upper types.Tuple) Range {
+	return Range{
+		Below{key: lower},
+		Above{key: upper},
+	}
+}
+
+// OpenClosedRange returns a range representing {l < x <= u}.
+func OpenClosedRange(lower, upper types.Tuple) Range {
+	return Range{
+		Above{key: lower},
+		Above{key: upper},
+	}
+}
+
+// ClosedOpenRange returns a range representing {l <= x < u}.
+func ClosedOpenRange(lower, upper types.Tuple) Range {
+	return Range{
+		Below{key: lower},
+		Below{key: upper},
+	}
+}
+
+// LessThanRange returns a range representing {x < u}.
+func LessThanRange(upper types.Tuple) Range {
+	return Range{
+		BelowAll{},
+		Below{key: upper},
+	}
+}
+
+// Returns a range representing  {x <= u}.
+func UpToRange(upper types.Tuple) Range {
+	return Range{
+		BelowAll{},
+		Above{key: upper},
+	}
+}
+
+// GreaterThanRange returns a range representing {x > l}.
+func GreaterThanRange(lower types.Tuple) Range {
+	return Range{
+		Above{key: lower},
+		AboveAll{},
+	}
+}
+
+// AtLeastRange returns a range representing {x >= l}.
+func AtLeastRange(lower types.Tuple) Range {
+	return Range{
+		Below{key: lower},
+		AboveAll{},
+	}
+}
+
+// AllRange returns a range representing all values.
+func AllRange() Range {
+	return Range{
+		BelowAll{},
+		AboveAll{},
+	}
+}
+
+// EmptyRange returns the empty range.
+func EmptyRange() Range {
+	return Range{
+		Above{},
```

Comment:
This will panic for Contains

Reply:
I think I can just remove `Contains` altogether